Assets/Scripts/Enemy.cs

```diff
@@ -2,34 +2,57 @@
 using System.Collections.Generic;
 using UnityEngine;
 
-public class Enemy : MonoBehaviour
-{
+/*
+ * Abstract class providing framework for enemy script design
+ */
+public abstract class Enemy : MonoBehaviour {
+    // Each enemy state has a corresponding behaviour the enemy performs while in that state
+    /* 
+     * TODO: Research if there's a more computationally efficient way to allow editing of EnemyBehaviour members in Inspector than serializing the EnemyBehaviours by reference.
+     * Example possibilities include defining a custom Editor subclass to add EnemyBehaviour members to Inspector.
+     */
+    [SerializeReference]
+    protected EnemyBehaviour _idleBehaviour, _awareBehaviour, _engagedBehaviour;
+    // We assume by default that every enemy is initially unaware of the player
+    protected ENEMY_STATE _state = ENEMY_STATE.idle;
+    // By initializing this as true, we ensure an enemy can change state on its first frame
+    protected bool _can_change_state = true;
 
-    private Rigidbody2D myRigidbody;
-    private SpriteRenderer mySprite;
-    public Rigidbody2D player;
-    // Start is called before the first frame update
-    public float max_speed = 4;
-    public float acceleration_rate = 1;
-    private void Awake(){
-        myRigidbody = GetComponent<Rigidbody2D>();
-        mySprite = GetComponent<SpriteRenderer>();
-    }
-    void Start()
-    {
-        player = GameObject.FindGameObjectWithTag("Player").GetComponent<Rigidbody2D>();
+    // We assume all enemies will have a Rigidbody2D for collision purposes
+    protected Rigidbody2D _myRigidbody;
+
+    // We enforce that an enemy define a Start method to initialize its enemy-specific behaviours
+    protected abstract void Start();
+
+    // We perform updates to enemies on FixedUpdate since they are physics objects
+    // FixedUpdate is virtual to allow overriding by custom non-state-based enemies (such as bosses, potentially)
+    protected virtual void FixedUpdate() {
+        if(this._can_change_state) {
+            this.ChangeState();
+        }
+        try {
+            this._can_change_state = this.PerformAction();
+        } catch(InvalidEnemyStateException ex) {
+            Debug.LogException(ex);
+            this._state = ENEMY_STATE.idle;
+            this._can_change_state = true;
+        }
```

Comment:
Is this really what we want to happen? Catching/handling exceptions implies we tolerate it and it is within the realm of expected behavior. Isn't this error indicative of something which is a relatively critical mistake in the backend that should be rectified directly, rather than tolerated and compensated for? Formally, shouldn't we allow this error to propagate and interrupt rather than catch and handle it directly